src/pip/_internal/resolution/resolvelib/requirements.py

```diff
@@ -33,8 +34,14 @@ def name(self):
         # No need to canonicalise - the candidate did this
         return self.candidate.name
 
-    def find_matches(self):
-        # type: () -> Sequence[Candidate]
+    def find_matches(self, constraint):
+        # type: (Sequence[SpecifierSet]) -> Sequence[Candidate]
+        if len(constraint) > 0:
+            raise InstallationError(
+                "Could not satisfy constraints for '{}': "
+                "installation from path or url cannot be "
+                "constrained to a version".format(self.name)
+            )
```

Comment:
Is this stale? The provider is now passing in `SpecifierSet`, not `Sequence[SpecifierSet]`. Also the `len()` part works? 🤔

Reply:
`SpecifierSet` implements `__len__` to give the number of specifiers in the set. See [here](https://packaging.pypa.io/en/latest/specifiers/#packaging.specifiers.SpecifierSet.__len__).

I was quite pleased when I found this 🙂